app/server/api-generator/src/main/java/io/syndesis/server/api/generator/openapi/OpenApiConnectorGenerator.java

```diff
@@ -136,11 +138,11 @@ public final APISummary info(final ConnectorTemplate connectorTemplate, final Co
     protected ConnectorDescriptor createDescriptor(String connectorId, OpenApiModelInfo info, OasOperation operation) {
         switch (info.getApiVersion()) {
```

Comment:
Would it make sense to have a field with `DataShapeGenerator<T,O>` and then two instances of `OpenApiConnectorGenerator` one with the field initialized to v2 implementation and one to v3 implementation. Would reduce the amount of conditionals we use here.
Then it would perhaps make sense to have the `OpenApiConnectorGenerator` parametrized with the same `<T extends OasDocument, O extends OasOperation>`.

Reply:
No, the `OpenApiConnectorGenerator` is a single entry point for other modules like `server-runtime`. Depends on if we want to have two separated server endpoints for the Syndesis UI to call (one for API v2 and one for API v3). Then we can separate the OpenAPI versions from the very beginning as the UI logic picks the right version. At the moment we have one API generating endpoint on the server.

Reply:
Perhaps I didn't make it that clear, this is what I was thinking:

```java

// from client, perhaps something like
OpenApiConnectorGenerator<?, ? generator = OpenApiConnectorGenerator.forSpecification(String specification);

// in the implementation
class OpenApiConnectorGenerator<T extends OasDocument, O extends OasOperation> {
  DataShapeGenerator<T,O> dataShapeGenerator;
  OpenApiDescriptorGenerator<T,O> descriptorGenerator;
  OpenApiPropertyGenerator<T, O> propertyGenerator;

  private static final OpenApiConnectorGenerator<Oas30Document, Oas30Operation> V3IMPL = new OpenApiConnectorGenerator(new Oas30DataShapeGenerator(), new Oas30DescriptorGenerator(), new Oas30PropertyGenerators());

  private OpenApiConnectorGenerator(DataShapeGenerator<T,O> generator, dataShapeGenerator, OpenApiDescriptorGenerator<T,O> descriptorGenerator,  OpenApiPropertyGenerator<T, O> propertyGenerator) {
    this.dataShapeGenerator = dataShapeGenerator;
    this.descriptorGenerator = descriptorGenerator;
    this.propertyGenerator = propertyGenerator;
  }

  public static <T extends OasDocument, O extends OasOperation> OpenApiConnectorGenerator<T, O> forSpecification(String specification) {
    JsonNode nodes = JsonUtils.reader().readTree(specification);
    OpenApiVersion openApiVersion = OpenApiModelParser.getOpenApiVersion(tree);

	if (OpenApiVersion == OpenApiVersion.V2) {
       // could be even cached in a static field
       return new OpenApiConnectorGenerator(new Oas20DataShapeGenerator(), new Oas20DescriptorGenerator(), new Oas20PropertyGenerators())
    } else if (OpenApiVersion == OpenApiVersion.V3) {
       // like here we cache it
       return V3IMPL;
    }
  }
}
```

Does that make sense, am I off base?

Reply:
Ah thanks, more clear now. And I like it! Will add it along with some upcoming work on that OpenAPI 3 support.